hips/hip-9999.md

```diff
@@ -0,0 +1,431 @@
+
+For the initial release of the new plugin system, chart authors will be able to define custom plugins for the following categories:
+
+- **Download** (to allow chart authors to specify plugins for downloading subcharts with different protocols than the default. See the `download` type under alternative behavior plugins)
+- **Manifest render** (for rendering a chart's templates into a set of Kubernetes resources using some other engine than gotemplate)
+
+After the initial plugin system release, we intend to make it easy to continue adding new chart-defined plugin types to extend additional categories of non-default chart behavior as this becomes desirable. Some examples may be:
+
+- Values schema validation (for validating the chart's `values.yaml` file using something other than JSON Schema)
+- Dependency resolution (for using a different subchart dependency resolver than the one currently used by Helm)
+
+To plan for forward compatibility, a `minimumHelmVersion` field may be added to allow future plugins to specify the minimum version of Helm that must be used for the chart (since this is the version of Helm that introduced the new plugin, perhaps this can be auto-detected). 
+
+Note that Helm 3 allowed a similar forwards-compatibility for Chart API v2, for example when support for a new gotemplate/sprig function was added to a specific Helm version. Charts could check the Helm version in their template or a helper function, and exit with a friendly error if the version of Helm did not meet this version constraint. We may be able to do better as this applies to plugins in Helm 4.
+
+A first pass for Chart-defined plugin syntax in `Chart.yaml`:
+
+```yaml
+# made up example URLs—these are not actual Helm 4 plugins yet
+plugins:
+    render:
+        - https://github.com/pkl-community/helm-pkl@0.1.0: [".pkl", ".pickle"]
+        - https://github.com/kubeys/helmys@0.1.4: [".ys"]
+    download:
+        - https://artifacthub.io/packages/helm-plugin/s3/s3@0.16.3: ["s3"]
+        - https://artifacthub.io/packages/helm-plugin/git/helm-git: ["git", "git+https", "git+ssh", "git+file"]
+```
+
+#### 4. CLI plugins
+
+The new plugin system will continue to allow users to build and install CLI plugins, which specify subcommands for the Helm CLI. However, these will also be run in Wasm sandbox rather than directly calling arbitrary executables as a subprocess as we done in Helm 3. This will require Helm 4 plugin developers to follow our tutorials for migrating their legacy CLI plugin to the new Wasm CLI plugin framework.
```

Comment:
```suggestion
The new plugin system will continue to allow users to build and install CLI plugins, which specify subcommands for the Helm CLI. However, these will also be run in Wasm sandbox rather than directly calling arbitrary executables as a subprocess as was implemented in Helm 3. This will require Helm 4 plugin developers to follow documentation which will be created detailing the process for migrating legacy CLI plugins to the new Wasm CLI plugin framework.
```